Add expected behavior to unified memory (#3617)
ROCm-Developer-Tools/HIP: docs/how-to/hip_runtime_api/memory_management/virtual_memory.rst

.. meta::
  :description: This chapter describes introduces Virtual Memory (VM) and shows
                how to use it in AMD HIP.
  :keywords: AMD, ROCm, HIP, CUDA, virtual memory, virtual, memory, UM, APU

.. _virtual_memory:

********************************************************************************
Virtual memory management
********************************************************************************

Memory management is important when creating high-performance applications in
the HIP ecosystem. Both allocating and copying memory can result in bottlenecks,
which can significantly impact performance.

Global memory allocation in HIP uses the C language style allocation function.

> This works fine for simple cases but can cause problems if your memory needs

Reviewer: ```suggestion
This works fine for simple cases but can cause problems if your application requires additional memory.
```

> change. If you need to increase the size of your memory, you must allocate a

Reviewer: ```suggestion
If you need to increase the size of your memory, you must allocate a
```

second larger buffer and copy the data to it before you can free the original
buffer. This increases overall memory usage and causes unnecessary ``memcpy``
calls. Another solution is to allocate a larger buffer than you initially need.
However, this isn't an efficient way to handle resources and doesn't solve the
issue of reallocation when the extra buffer runs out.

> Virtual memory management solves these memory management problems. It helps to

Reviewer: ```suggestion
Virtual memory management solves these problems by reducing
```

> reduce memory usage and unnecessary ``memcpy`` calls.

Reviewer: ```suggestion
memory usage and unnecessary ``memcpy`` calls.
```

================================================================================

Standard memory allocation uses the :cpp:func:`hipMalloc` function to allocate a
block of memory on the device. However, when using virtual memory, this process
is separated into multiple steps using the :cpp:func:`hipMemCreate`,
:cpp:func:`hipMemAddressReserve`, :cpp:func:`hipMemMap`, and
:cpp:func:`hipMemSetAccess` functions. This guide explains what these functions
do and how you can use them for virtual memory management.

Allocate physical memory
--------------------------------------------------------------------------------

The first step is to allocate the physical memory itself with the
:cpp:func:`hipMemCreate` function. This function accepts the size of the buffer,
an ``unsigned long long`` variable for the flags, and a
:cpp:struct:`hipMemAllocationProp` variable. :cpp:struct:`hipMemAllocationProp`
contains the properties of the memory to be allocated, such as where the memory
is physically located and what kind of shareable handles are available. If the
allocation is successful, the function returns a value of
:cpp:enumerator:`hipSuccess`, with :cpp:type:`hipMemGenericAllocationHandle_t`

> representing a valid physical memory allocation. The allocated memory size must

Reviewer: ```suggestion
representing a valid physical memory allocation. 

The allocated memory size must
```

I would at least separate these last two sentences out of the preceding paragraph. I might also suggest making them a Note perhaps? Finally, I think it needs to be more definite about who needs to align the memory size with the granularity... _You_ must align the memory size...

    hipMemCreate(&allocHandle, padded_size, &prop, 0);

Reserve virtual address range
--------------------------------------------------------------------------------

After you have acquired an allocation of physical memory, you must map it before
you can use it. To do so, you need a virtual address to map it to.  Mapping
means the physical memory allocation is available from the virtual address range
it is mapped to. To reserve a virtual memory range, use the
:cpp:func:`hipMemAddressReserve` function. The size of the virtual memory must
match the amount of physical memory previously allocated. You can then map the
physical memory allocation to the newly-acquired virtual memory address range
using the :cpp:func:`hipMemMap` function.

.. code-block:: cpp

    hipMemAddressReserve(&ptr, padded_size, 0, 0, 0);
    hipMemMap(ptr, padded_size, 0, allocHandle, 0);

Set memory access
--------------------------------------------------------------------------------

Finally, use the :cpp:func:`hipMemSetAccess` function to enable memory access.
It accepts the pointer to the virtual memory, the size, and a
:cpp:struct:`hipMemAccessDesc` descriptor as parameters. In a multi-GPU

> environment, you can map the device memory of one GPU to another. This feature

Reviewer: ```suggestion
environment, you can map the device memory of one GPU to another. 

This feature
```
 would start a new paragraph with "This feature..."

between the devices. When using virtual memory management, peer access is
enabled by :cpp:func:`hipMemSetAccess`, which provides a finer level of
control over what is shared. This has no performance impact on memory allocation
and gives you more control over what memory buffers are shared with which
devices.

.. code-block:: cpp

    hipMemAccessDesc accessDesc = {};
    accessDesc.location.type = HIP_MEM_LOCATION_TYPE_DEVICE;
    accessDesc.location.id = currentDev;
    accessDesc.flags = HIP_MEM_ACCESS_FLAGS_PROT_READWRITE;
    hipMemSetAccess(ptr, padded_size, &accessDesc, 1);

At this point the memory is allocated, mapped, and ready for use. You can read
and write to it, just like you would a C style memory allocation.

Free virtual memory
--------------------------------------------------------------------------------

To free the memory allocated in this manner, use the corresponding free
functions. To unmap the memory, use :cpp:func:`hipMemUnmap`. To release the
virtual address range, use :cpp:func:`hipMemAddressFree`.  Finally, to release
the physical memory, use :cpp:func:`hipMemRelease`. A side effect of these

> functions is the lack of synchronization when memory is released. If you call

Reviewer: If you call :cpp:func:`hipFree` when you have multiple streams running in parallel, it
synchronizes the device. This causes worse resource usage and performance.

This seems like it could be an important note?

:cpp:func:`hipFree` when you have multiple streams running in parallel, it
synchronizes the device. This causes worse resource usage and performance.

.. code-block:: cpp

    hipMemUnmap(ptr, size);
    hipMemRelease(allocHandle);
    hipMemAddressFree(ptr, size);

.. _usage_virtual_memory:

Memory usage
================================================================================

Dynamically increase allocation size
--------------------------------------------------------------------------------

The :cpp:func:`hipMemAddressReserve` function allows you to increase the amount
of pre-allocated memory. This function accepts a parameter representing the

> requested starting address of the virtual memory. This allows you to have a

Reviewer: ```suggestion
requested starting address of the virtual memory, and the amount of additional memory to reserve. This allows you to have a
```
